File: occurs.lua
#!/usr/bin/env lua
spec = [=[
occurs 0.92
Copyright (c) 2022 Reuben Thomas <rrt@sc3d.org>

Usage: occurs [OPTION...] [FILE...]

Count the occurrences of each symbol in input.

The default symbol type is words (-s "[[:alpha:]]+"); other useful settings
include:

  non-white-space characters: -s "[^[:space:]]+"
  alphanumerics and underscores: -s "[[:alnum:]_]+"
  XML tags: -s "<([a-zA-Z_:][a-zA-Z_:.0-9-]*)[[:space:]>]"

Options:

  -s, --symbol=REGEXP      symbols are given by REGEXP
  -n, --nocount            don't show the frequencies or total
      --help               display this help, then exit
      --version            display version information, then exit
]=]

local std = require "std"
rex_posix = require "rex_posix"

-- Parse command-line args
os.setlocale ("")
local OptionParser = require "std.optparse"
local parser = OptionParser (spec)
_G.arg, opts = parser:parse (_G.arg)
local symbolPat = opts.symbol or "[[:alpha:]]+"

-- Compile symbol-matching regexp
local ok, pattern = pcall (rex_posix.new, symbolPat)
if not ok then
  die (pattern)
end

-- Process input
local freq = {}
std.io.process_files (function (file, number)
                        -- FIXME: make slurp work in pipes
                        for s in rex_posix.gmatch (std.io.slurp (), pattern) do
                          freq[s] = (freq[s] or 0) + 1
                        end
                      end)

-- Write output
local symbols = 0
for s in pairs (freq) do
  std.io.writelines (s .. (opts.nocount and "" or " " .. freq[s]))
  symbols = symbols + 1
end
if not opts.nocount then
  std.io.warn ("total symbols: " .. symbols)
end

File: occurs.nim
import os
import system
import argparse
import re
import logging
import strformat
import std/tables


const progName = "occurs" # FIXME: allow newParser to take a variable, then use: lastPathPart(getAppFilename())

# Error messages
var logger = newConsoleLogger(fmtStr = &"{progName}: ")
addHandler(logger)
proc die(s: string) =
  warn(s)
  quit(1)

# Command-line arguments
var p = newParser(progName):
  help("Count the occurrences of each symbol in input.")
  option("-s", "--symbol", default = "\\p{L}+",
         help = "symbols are given by the regexp SYMBOL")
  flag("-n", "--nocount",
         help = "don't show the frequencies or total")
  flag("-V", "--version",
       help = "show program's version number and exit")
  arg("file", nargs = -1)
  nohelpflag()
  flag("-h", "--help", help = "show this help") # FIXME: downcase help message for --help

try:
  let opts = p.parse()
  if opts.help:
    echo """

The default symbol type is words (-s "\p{L}+"); other useful settings
include:

  non-white-space characters: -s "\S+"
  alphanumerics and underscores: -s "\w+"
  XML tags: -s "<([a-zA-Z_:][a-zA-Z_:.0-9-]*)[\s>]""""
    quit(0)
  if opts.version:
    echo &"{progName} 0.11 (16 Sep 2022) by Reuben Thomas <rrt@sc3d.org>"
    quit(0)

  # Compile symbol-matching regexp
  let pattern = re(opts.symbol)

  # Process input
  var freq = initOrderedTable[string, int]()
  for f in opts.file #[or ['-']]#:
    for line in f.lines:
      for s in findAll(line, pattern):
        if s notin freq:
          freq[s] = 0
        freq[s] += 1

  # Write output
  var symbols = 0
  for s, f in pairs(freq):
    let freq_output = if opts.nocount: "" else: &" {f}"
    echo &"{s}{freq_output}"
    symbols += 1
  if not opts.nocount:
    warn(&"total symbols: {symbols}")

except: die(getCurrentExceptionMsg())
